### ISSUE TITLE
500 error when trying to log in.

### ISSUE THREAD

**Issue by curtislacy** [errbit/errbit#969]
Mongolab recently moved our database to mongoDB 3, and I updated errbit to the current master branch today.  When errors come in, they're pushed out to Hipchat, and that part works.  However, when I try to actually log into errbit, I get this message in my browser:

<img width="429" alt="screen shot 2015-10-14 at 4 09 47 pm" src="https://cloud.githubusercontent.com/assets/1965022/10496300/0d695b5a-728e-11e5-8be6-50c4e02c32da.png">

And this one in the log (returned by ```heroku logs```:

```
2015-10-14T19:47:23.117890+00:00 app[web.1]: 
2015-10-14T19:47:23.117893+00:00 app[web.1]: ActionView::Template::Error (uninitialized constant IssueTrackers):
2015-10-14T19:47:23.117895+00:00 app[web.1]:     10:   %thead
2015-10-14T19:47:23.117895+00:00 app[web.1]:     11:     %tr
2015-10-14T19:47:23.117896+00:00 app[web.1]:     12:       %th= t('.name')
2015-10-14T19:47:23.117898+00:00 app[web.1]:     14:         %th= t('.repository')
2015-10-14T19:47:23.117897+00:00 app[web.1]:     13:       - if any_github_repos? || any_bitbucket_repos?
2015-10-14T19:47:23.117898+00:00 app[web.1]:     15:       - if any_notification_services?
2015-10-14T19:47:23.117899+00:00 app[web.1]:     16:         %th= t('.notify')
2015-10-14T19:47:23.117900+00:00 app[web.1]:   app/models/app.rb:138:in `issue_tracker_configured?'
2015-10-14T19:47:23.117901+00:00 app[web.1]:   app/helpers/apps_helper.rb:47:in `block in detect_any_apps_with_attributes'
2015-10-14T19:47:23.117901+00:00 app[web.1]:   app/helpers/apps_helper.rb:44:in `each'
2015-10-14T19:47:23.117902+00:00 app[web.1]:   app/helpers/apps_helper.rb:44:in `detect_any_apps_with_attributes'
2015-10-14T19:47:23.117904+00:00 app[web.1]:   app/views/apps/index.html.haml:13:in `_app_views_apps_index_html_haml__3335945173520495906_70049067902880'
2015-10-14T19:47:23.117903+00:00 app[web.1]:   app/helpers/apps_helper.rb:15:in `any_github_repos?'
```

Not sure if the error above should be fatal, or only crashes everything because of https://github.com/errbit/errbit/issues/968

This is running mongod version: 3.0.6 on mongolab.
I believe I'm up to date on migrations.  ```heroku run rake db:version``` returns:

```
Running rake db:version on dtt-errbit... up, run.4165
Notice: no rspec tasks available in this environment
Overwriting existing field _id in class App.
Creating scope :page. Overwriting existing method NotificationServices::CampfireService.page.
Creating scope :page. Overwriting existing method NotificationServices::FlowdockService.page.
Creating scope :page. Overwriting existing method NotificationServices::GtalkService.page.
Creating scope :page. Overwriting existing method NotificationServices::HipchatService.page.
Creating scope :page. Overwriting existing method NotificationServices::HoiioService.page.
Creating scope :page. Overwriting existing method NotificationServices::HubotService.page.
Creating scope :page. Overwriting existing method NotificationServices::PushoverService.page.
Creating scope :page. Overwriting existing method NotificationServices::SlackService.page.
Creating scope :page. Overwriting existing method NotificationServices::WebhookService.page.
20150926035420
```

**Comment by stevecrozz** [comment]
@curtislacy Thanks for the report. In all cases where I've seen this error so far, the answer has been that this migration was not fully applied:
https://github.com/errbit/errbit/blob/v0.4.0/db/migrate/20131011155638_extract_issue_tracker.rb

If you have a log from when this migration was applied, I'd love to see it because I know a few others have run into this, but I haven't yet figured out why it happens.

**Comment by curtislacy** [comment]
@stevecrozz Thanks for the reply - I'd seen that migration mentioned in a few other places as I looked into this, and I'm inclined to believe you.  I'm not sure exactly when it was applied though, so digging up the log for it is tricky.  Is there a way to roll down the migrations until this one is unapplied, and then reapply it?  Or does the empty ```self.down``` function (https://github.com/errbit/errbit/blob/v0.4.0/db/migrate/20131011155638_extract_issue_tracker.rb#L44-L45) make that as much of a non-starter as I fear?

**Comment by stevecrozz** [comment]
@curtislacy It's probably better to apply it by hand by running it in ./bin/rails console once we have an idea of what's up. It probably could be run backwards and back up as most of the migrations are idempotent, but I don't think the migration framework allows for doing this on a single migration in history. I think they are applied as a chain, so rolling back would involve rolling back any migrations in between as well.

But not to worry. I'm confident we can fix this Errbit deployment, and I'm hopeful we can even find the cause if you can find the logs so others don't run into it. If you want to jump on https://gitter.im/errbit/errbit I can help dig through it with you.

**Comment by jerefrer** [comment]
Okay I ran into the same issue for the same reason (mongo 3 update on mongolab).

Applying by hand to migration worked for me, but I had to fix `App.collection.where` to `App.where`.

You'll find the whole script to run below:

```ruby
TRACKER_MAPPING = {
  'ErrbitTracPlugin::IssueTracker' => 'trac',
  'IssueTrackers::BitbucketIssuesTracker' => 'bitbucket',
  'IssueTrackers::FogbugzTracker' => 'fogbugz',
  'IssueTrackers::GithubIssuesTracker' => 'github',
  'IssueTrackers::GitlabTracker' => 'gitlab',
  'IssueTrackers::JiraTracker' => 'jira',
  'IssueTrackers::LighthouseTracker' => 'lighthouse',
  'IssueTrackers::PivotalLabsTracker' => 'pivotal',
  'IssueTrackers::RedmineTracker' => 'redmine',
  'IssueTrackers::UnfuddleTracker' => 'unfuddle'
}

App.all.each do |app|
  next unless app.attributes['issue_tracker'].present?
  next unless app.attributes['issue_tracker']['_type'].present?

  options = app['issue_tracker'].dup
  options.delete('_type')
  options.delete('_id')

  _type = app.attributes['issue_tracker']['_type']
  updated_at = options.delete('updated_at')
  created_at = options.delete('created_at')

  if TRACKER_MAPPING.include?(_type)
    tracker = {
      'type_tracker' => TRACKER_MAPPING[_type],
      'options' => options,
      'updated_at' => updated_at,
      'created_at' => created_at
    }

    App.where({ _id: app.id }).update({
      "$set" => { :issue_tracker => tracker }
    })
  end
end
```

**Comment by stevecrozz** [comment]
I'm going to close this issue because it really just seems like this old migration was never applied, but I want to leave a bit more info here in case anyone else is running into this.

v0.5.0 and master errbit don't contain the old migrations, so if these old migrations were never run, that cannot be fixed automatically in v0.5.0+ since I purged all the old migrations from the codebase. The answer is to upgrade first to v0.3.0 or v0.4.0, run migrations, then upgrade to v0.5.0 and run them again.

If you found this page and you're experiencing this same issue while following [these instructions](https://github.com/errbit/errbit#upgrading-errbit-beyond-v040) then I'm wrong about how this happens and we should reopen this, especially if you have a log file that can show what went wrong at the time of the extract_issue_tracker migration.
